Hotfix cache queries for performance

CarlosQ96 opened the pull request:
Used Threads Library to off load hashing of cache keys.
Added cache for count and results of projects filters queries.
Moved main filters query to projectRepository (this is because queries share state, so instantiating different query builders avoid wrong results when using cache).

CarlosQ96 committed:
separate queries and run them parallel, add cache to categories

CarlosQ96 committed:
update allProjects query with cache and separated queries

CarlosQ96 committed:
add cache key generation logic

CarlosQ96 committed:
add types import to cache key generation

CarlosQ96 committed:
install threads npm package

CarlosQ96 committed:
delete old projects filters graphql query

CarlosQ96 committed:
add threads to run heavy processes

CarlosQ96 committed:
add thread pools for workers

CarlosQ96 committed:
fix tests to support cache

CarlosQ96 committed:
add cache env vars to example.env

CarlosQ96 committed:
fix pool spawning when running server

CarlosQ96 commented:
To debug you can add ```DEBUG=threads:*``` to any command like npm start or npm run test

CarlosQ96 committed:
fix thread pool instantiation

CarlosQ96 commented on src/repositories/projectRepository.ts:
```diff
@@ -28,6 +38,85 @@ export const findProjectById = (
     .getOne();
 };
 
+// return query without execution
```
Moved Filter query to the repository so I could reuse it for count query and results query.
When using the same querybuilder object, the cache was throwing errors.

CarlosQ96 commented on src/server/bootstrap.ts:
```diff
@@ -135,6 +148,12 @@ export async function bootstrap() {
       }
     }
 
+    // instantiate pool once and pass as context
```
If you separate it on another module ```export const pool``` throws tons of errors.
If you put it at the root of the project resolver, thread instantiation gets stuck.

This is the only way I found it worked. Instantiate it in the express server and pass it as context to the graphql resolvers.

aminlatifi commented on src/utils/utils.ts:
```diff
@@ -1,9 +1,32 @@
+}) => {
+  const orderedArgs = stringify(args);
+
+  return await hashMD5(orderedArgs, { algorithm: 'md5' });
```
`object-hash` has the parameter
```
unorderedObjects {true|false} Sort objects before hashing, i.e. make hash({ x: 1, y: 2 }) === hash({ y: 2, x: 1 }). default: true
```
As mentioned, its default value is true. As we order and stringify by `json-stable-stringify`, it's not efficient to order the object twice. We can either set that parameter as false, or skip `json-stable-stringify`

CarlosQ96 replied:
Ok awesome, I missed that detail, I think in this case we should remove the json-stringify, less packages in our project the better.

aminlatifi commented on src/server/bootstrap.ts:
```diff
@@ -66,6 +67,18 @@ const cors = require('cors');
+    process.env.PROJECT_FILTERS_THREADS_POOL_CONCURRENCY || 1,
+  ),
+  maxQueuedJobs: Number(
+    process.env.PROJECT_FILTERS_THREADS_POOL_MAX_QUEUED || 4,
```
Why do we override its default unlimited value? If we get many queries simultaneously it will cause issue.

CarlosQ96 replied:
I was probably thinking unlimited stuff is bad hahaha

but I will remove the limit as we don't want now to break the query when we have high loads of people, we want performance not new issues.

aminlatifi commented on src/resolvers/projectResolver.ts:
```diff
@@ -567,221 +573,88 @@ export class ProjectResolver {
-      );
+      user,
+    );
+    const filtersResultHashTask = projectsFiltersThreadPool.queue(hasher =>
```
Why we run two queries, one for count and one for fetching the projects?

CarlosQ96 replied:
I will test GetManyAndCount() query with 1 cache and let you know. I think I avoided it because I was getting errors with the cache. Will report back on this comment.

aminlatifi commented on src/resolvers/projectResolver.ts:
```diff
@@ -567,221 +573,88 @@ export class ProjectResolver {
+      filters,
+      sortingBy,
+      connectedWalletUserId,
+      user,
```
I think we must skip the user from the key. I mean fetch the projects independent of the user, then in a separate query load the user reactions and fill projects! Of course, db does the joining better than us, but 
- For every user cache value is useless
- User may update the reaction (like/unlike a project)! and the cache will not be invalidated immediately

CarlosQ96 replied:
As we decided by DM, I will remove the reaction of the logged in user portion from the cache and the userId from the cache key. 
Query the reaction separately and then in a Worker merge the data.

CarlosQ96 committed:
remove limit of thread queues, remove json-stable-stringify, simplify query verbosity

CarlosQ96 committed:
simplify hashing of filter params

CarlosQ96 committed:
merge user Reactions with projects result and simplify allProjects Query

CarlosQ96 committed:
Merge branch 'staging' into hotfix_cache_queries_for_performance

aminlatifi committed:
Improved merge user reaction to projects
Used json-stable-stringify and nodejs native hashing instead of object-hash library
Removed user related data from computing allProject query hash

aminlatifi commented:
Made[ this improvement PR](https://github.com/Giveth/impact-graph/pull/810)

We have a build issue with typescript v3.
https://github.com/andywer/threads.js/issues/359#issuecomment-831966607
So we are blocked by other dependency upgrade PRs #808 and #630

aminlatifi committed:
Merge pull request #810 from Giveth/hotfix_cache_queries_for_performance_improve

improve cache queries

aminlatifi commented:
Merged https://github.com/Giveth/impact-graph/pull/810 as tests passed.
The major idea behind improvements:
1. Merging reaction complexity improved, from O(n^2) to O(n)
2. Relay on json-stable-stringify, since it support nested object sorting. object-hash does only first level key sorting

aminlatifi committed:
Added type to project resolver module thread

aminlatifi committed:
Merge branch 'staging' into hotfix_cache_queries_for_performance

# Conflicts:
#	config/test.env
#	package-lock.json
#	package.json
#	src/repositories/projectRepository.ts
#	src/repositories/reactionRepository.ts
#	src/resolvers/projectResolver.test.ts
#	src/resolvers/projectResolver.ts
#	src/server/bootstrap.ts

aminlatifi committed:
Removed unused imports

aminlatifi committed:
Removed a test function wrongly inserted on conflict resolving

aminlatifi committed:
Merge branch 'staging' into hotfix_cache_queries_for_performance

aminlatifi commented:
@CarlosQ96 I merged staging into it, and everything seems okay.
Please resolve the above comments, and then we can merge this. Do you think we need to configure the staging env? or the default values are fine?

CarlosQ96 commented:
Hey Amin for staging I think the default values are enough.
